### PR TITLE
CI and testing infrastructure improvements.

### PR DESCRIPTION
**Pull request by zaneselvans**
# Overview

- Update the `setup-micromamba` action to v2
- Update the `micromamba` docker image to use the recently released 1.5.10 tag
- Update to v4.0.0 of `pre-commit` and migrate config file
- Stop using paid large runner for `pytest-integration` since open source projects get 4vCPU / 16GB runners for free [as of January, 2024](https://github.blog/news-insights/product-news/github-hosted-runners-double-the-power-for-open-source/)
- Fix a bug in our Makefile where `make install-pudl` would fail if the `pudl-dev` environment didn't already exist.
- Relock conda dependencies.

# Testing

## How did you make sure this worked? How can a reviewer verify this?

```[tasklist]
# To-do list
- [x] Run `make pytest-coverage` locally to ensure that the merge queue will accept your PR.
- [x] Review the PR yourself and call out any questions or issues you have.
```

### CODE REVIEW

**Comment by zaneselvans** [comment]
See [this pytest action run](https://github.com/catalyst-cooperative/pudl/actions/runs/11204094947) for confirmation that the integration tests work without the big runner.

File: .github/workflows/pytest.yml [hunk 1]
```diff
@@ -120,9 +120,7 @@ jobs:
           path: .coverage
 
   ci-integration:
-    runs-on:
-      group: large-runner-group
-      labels: ubuntu-latest-4core
+    runs-on: ubuntu-latest
```
Comment (zaneselvans) [inline .github/workflows/pytest.yml:123]: See [this pytest action run](https://github.com/catalyst-cooperative/pudl/actions/runs/11204094947) for confirmation that the larger paid runner is no longer necessary.

File: .github/actionlint.yaml
Comment (zaneselvans) [inline .github/actionlint.yaml:1]: Configuration no longer required because we don't need to use a special large runner.

File: Makefile [hunk 2]
```diff
@@ -52,7 +52,10 @@ conda-lock.yml: pyproject.toml
 .PHONY: pudl-dev
 pudl-dev:
 	${mamba} run --name base ${mamba} install --quiet --yes "conda-lock>=2.5.7"
-	${mamba} run --name base ${mamba} env remove --yes --name pudl-dev
+# Only attempt to remove the pudl-dev environment if it already exists.
+	if ${mamba} env list | grep -q pudl-dev; then \
+		${mamba} env remove --quiet --yes --name pudl-dev; \
+	fi
```
Comment (zaneselvans) [inline Makefile:58]: I discovered this was failing after removing my local `pudl-dev` environment, and attempting to recreate it with the make target. Oops.

File: docker/Dockerfile [hunk 3]
```diff
@@ -1,4 +1,4 @@
-FROM mambaorg/micromamba:1.5.9
+FROM mambaorg/micromamba:1.5.10
```
Comment (zaneselvans) [inline docker/Dockerfile:1]: As of yet there's no micromamba 2.0 image, but I imagine it'll come out shortly.

File: .github/actionlint.yaml
Comment (jdangerx) [reply .github/actionlint.yaml:1]: Yay!!

**Review by jdangerx — APPROVED** [review]
Glad you caught this, ha! Now to submit CI runs willy-nilly to make up for all that extra money we paid :)

**Pull request merged**